### PR TITLE
Enable config file usage and flags

### LINKED ISSUE #647: Load and use file-based configuration
See #625

### PR DESCRIPTION
**Pull request by theganyo**
Enables config file usage and flags as follows:

Flags enabled for registry CLI:

```
  -c, --config string             Name of a settings profile or path to config file
      --registry.address string   the server and port of the registry api (eg. localhost:8080)
      --registry.insecure         set to true if client must connect via http (not https)
      --registry.token string     the token to use for authorization to registry
```

Notes:

* This is 100% backward compatible
* `go test` will use env vars, not flags
* Settings are honored in prioritized order from: 1. flags, 2. env vars, 3. config file
* Config files are assumed to be stored in $HOME/.config/registry
* However, `--config` flag can specify an explicit path to a settings file directly

Config file management will be handled in #649 

Fixes #647

### CODE REVIEW

**Comment by timburks** [comment]
Let's watch out for possible collisions between the `--registry.*` flags and options in autogenerated commands.
```
$ registry rpc update-api --help
UpdateApi can be used to modify a specified API.

Usage:
  registry rpc update-api [flags]

Flags:
      --allow_missing                       If set to true, and the api is not found, a new...
      --api.annotations stringArray         key=value pairs. Annotations attach non-identifying metadata to...
      --api.availability string             A user-definable description of the availability...
      --api.description string              A detailed description.
      --api.display_name string             Human-meaningful name.
      --api.labels stringArray              key=value pairs. Labels attach identifying metadata to resources....
      --api.name string                     Resource name.
      --api.recommended_deployment string   The recommended deployment of the API.  Format:...
      --api.recommended_version string      The recommended version of the API.  Format:...
      --from_file string                    Absolute path to JSON file containing request payload
  -h, --help                                help for update-api
      --update_mask.paths strings           The set of field mask paths.

Global Flags:
  -j, --json      Print JSON output
  -v, --verbose   Print verbose output
```

I think we're ok now because none of our API methods are taking a `registry` message as a request argument.

**Comment by theganyo** [comment]
Hmm. I don't know how we could ensure that we don't collide with the current autogen scheme since it does't have it's own namespace...

**Comment by timburks** [comment]
I think the current option names are good. Just wanted to point out the possibility of future surprises.

File: pkg/connection/settings.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	home, err := os.UserHomeDir()
+	if err != nil {
+		panic(err)
+	}
+	configPath = filepath.Join(home, ".config/registry")
+
+	Flags = pflag.NewFlagSet("registry", pflag.ExitOnError)
+	Flags.StringP("config", "c", "", "Name of a settings profile or path to config file")
+	Flags.String("registry.address", "", "the server and port of the registry api (eg. localhost:8080)")
+	Flags.Bool("registry.insecure", false, "set to true if client must connect via http (not https)")
```
Comment (timburks) [inline pkg/connection/settings.go:46]: could we clarify this, possibly with "if specified, client connects via http (not https)"

**Review by timburks — COMMENTED** [review]
This is great!

We're going to need to add some more special handling for the generated RPC handlers. These two client creations are in generated code that will not use the new configuration.  I think it might be best to just patch these two functions, replacing their entire contents with new code that generates AdminClient and RegistryClient using `pkg/connection`. 

https://github.com/apigee/registry/blob/7bc06ba752571bcb29b327c4351e0a810cc9e283/cmd/registry/cmd/rpc/generated/admin_service.go#L76

https://github.com/apigee/registry/blob/7bc06ba752571bcb29b327c4351e0a810cc9e283/cmd/registry/cmd/rpc/generated/registry_service.go#L103

Other patching is done [here](https://github.com/apigee/registry/blob/main/tools/GENERATE-CLI.sh).

File: pkg/connection/settings.go (hunk 1 above)
Comment (theganyo) [reply pkg/connection/settings.go:46]: yup

**Comment by theganyo** [comment]
The commit I just pushed fixes the generated rpc cmds to use the same files and flags as the main registry cmd. However, it leaves quite a bit of dead code in the generate clients that we need to work on removing. The following is a diff example of what we have yet to do: https://gist.github.com/theganyo/99150cbc129557d991205c87528aa972

**Review by timburks — APPROVED** [review]
Looks great!

I'm ok with the patch code as-is. It will be easier to simplify it after #660 is merged - then I think all the patches could be done with the same set of (parameterized) steps.

File: pkg/connection/settings.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+// `cmd.PersistentFlags().AddFlagSet(connection.Flags)`
+var Flags *pflag.FlagSet
+
+var configPath string
+var activeConfigPointerFilename = "active_config"
+
+func init() {
+	home, err := os.UserHomeDir()
+	if err != nil {
+		panic(err)
```
Comment (timburks) [inline pkg/connection/settings.go:39]: nit: I think there's something in the style guide about panics, but it seems pretty hard to get an error here: https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/os/file.go;l=500

**Pull request merged**